qiskit/circuit/library/fourier_checking.py

```diff
@@ -72,26 +78,62 @@ def __init__(self, f: List[int], g: List[int]) -> None:
+        circuit = fourier_checking(f, g)
+        super().__init__(*circuit.qregs, name=circuit.name)
+        self.compose(circuit.to_gate(), qubits=self.qubits, inplace=True)
+
+
+def fourier_checking(f: List[int], g: List[int]) -> QuantumCircuit:
+    """Fourier checking circuit.
+
+    The circuit for the Fourier checking algorithm, introduced in [1],
+    involves a layer of Hadamards, the function :math:`f`, another layer of
+    Hadamards, the function :math:`g`, followed by a final layer of Hadamards.
+    The functions :math:`f` and :math:`g` are classical functions realized
+    as phase oracles (diagonal operators with {-1, 1} on the diagonal).
+
+    The probability of observing the all-zeros string is :math:`p(f,g)`.
+    The algorithm solves the promise Fourier checking problem,
+    which decides if f is correlated with the Fourier transform
+    of g, by testing if :math:`p(f,g) <= 0.01` or :math:`p(f,g) >= 0.05`,
+    promised that one or the other of these is true.
 
-        if len(f) != len(g) or num_qubits == 0 or not num_qubits.is_integer():
-            raise CircuitError(
-                "The functions f and g must be given as truth "
-                "tables, each as a list of 2**n entries of "
-                "{1, -1}."
-            )
+    The functions :math:`f` and :math:`g` are currently implemented
+    from their truth tables but could be represented concisely and
+    implemented efficiently for special classes of functions.
 
-        circuit = QuantumCircuit(num_qubits, name=f"fc: {f}, {g}")
+    Fourier checking is a special case of :math:`k`-fold forrelation [2].
 
-        circuit.h(circuit.qubits)
+    **Reference Circuit:**
 
-        circuit.compose(Diagonal(f), inplace=True)
+    .. plot::
 
-        circuit.h(circuit.qubits)
+       from qiskit.circuit.library import fourier_checking
+       circuit = fourier_checking([1, -1, -1, -1], [1, 1, -1, -1])
+       circuit.draw('mpl')
 
-        circuit.compose(Diagonal(g), inplace=True)
+    **Reference:**
 
-        circuit.h(circuit.qubits)
+    [1] S. Aaronson, BQP and the Polynomial Hierarchy, 2009 (Section 3.2).
+    `arXiv:0910.4698 <https://arxiv.org/abs/0910.4698>`_
 
-        super().__init__(*circuit.qregs, name=circuit.name)
-        self.compose(circuit.to_gate(), qubits=self.qubits, inplace=True)
+    [2] S. Aaronson, A. Ambainis, Forrelation: a problem that
+    optimally separates quantum from classical computing, 2014.
+    `arXiv:1411.5729 <https://arxiv.org/abs/1411.5729>`_
+    """
+    num_qubits = math.log2(len(f))
+
+    if len(f) != len(g) or num_qubits == 0 or not num_qubits.is_integer():
+        raise CircuitError(
+            "The functions f and g must be given as truth "
+            "tables, each as a list of 2**n entries of "
+            "{1, -1}."
+        )
+
+    circuit = QuantumCircuit(int(num_qubits), name=f"fc: {f}, {g}")
+    circuit.h(circuit.qubits)
+    circuit.compose(Diagonal(f), inplace=True)
```

Comment:
Can we directly use `DiagonalGate` here? `Diagonal` is a circuit that just uses `DiagonalGate` internally 🙂

Reply:
Great suggestion! Fixed. However, because this changes how the circuit looks like, I have reverted to use the old code for `FourierChecking` circuit.